store/decoupled/store.go

```diff
@@ -0,0 +1,212 @@
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/iavl"
+	"github.com/cosmos/cosmos-sdk/store/tracekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+	// "github.com/cosmos/cosmos-sdk/telemetry"
+)
+
+const (
+	defaultIAVLCacheSize = 10000
+)
+
+var (
+	_ types.KVStore                 = (*Store)(nil)
+	_ types.CommitStore             = (*Store)(nil)
+	_ types.CommitKVStore           = (*Store)(nil)
+	_ types.Queryable               = (*Store)(nil)
+	_ types.StoreWithInitialVersion = (*Store)(nil)
+)
+
+var (
+	versionsPrefix = []byte{0}
+	dataPrefix     = []byte{1}
+	indexPrefix    = []byte{2}
+)
+
+// A store which uses separate data structures for state storage (SS) and state commitments (SC)
+// TODO:
+// DB interface and SC store must support versioning
+type Store struct {
+	// Direct KV mapping (SS)
+	data dbm.DB
+	// Inverted index of SC values to SS keys
+	inv dbm.DB
+	// State commitments layer
+	sc types.CommitKVStore
+	// Mutex needed to lock stores in tandem during writes
+	mtx sync.Mutex
+}
+
+// TODO:
+// version tracking
+// constructors?
+// separate backing DBs for SC/SS?
+
+// Create a new, empty store
+func NewStore(db dbm.DB) (*Store, error) {
+	// tree, err := iavl.NewMutableTree(db, defaultIAVLCacheSize)
+	sc, err := iavl.LoadStore(db, types.CommitID{}, false)
+	if err != nil {
+		return nil, err
+	}
+	return newStore(db, sc), nil
+}
+
+// TODO: review logic, motivation for initial version - desired for this store? lazy loading?
+func LoadStore(db dbm.DB, id types.CommitID, initialVersion uint64) (types.CommitKVStore, error) {
+	sc, err := iavl.LoadStoreWithInitialVersion(db, id, false, initialVersion)
+	if err != nil {
+		return nil, err
+	}
+	return newStore(db, sc), nil
+}
+
+// Create a new store from SC store and DB
+func newStore(db dbm.DB, sc types.CommitKVStore) *Store {
+	return &Store{
+		sc:   sc,
+		data: dbm.NewPrefixDB(db, dataPrefix),
+		inv:  dbm.NewPrefixDB(db, indexPrefix),
+	}
+}
+
+// implement KVStore
+func (s *Store) Get(key []byte) []byte {
+	val, err := s.data.Get(key)
+	if err != nil {
+		panic(err)
+	}
+	return val
+}
+
+func (s *Store) Has(key []byte) bool {
+	has, err := s.data.Has(key)
+	return err == nil && has
+}
+
+func (s *Store) Set(key []byte, value []byte) {
+	kvHash := sha256.Sum256(append(key, value...))
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	err := s.data.Set(key, value)
+	if err != nil {
+		panic(err.Error())
+	}
+	err = s.inv.Set(kvHash[:], key)
+	if err != nil {
+		panic(err.Error())
+	}
+	s.sc.Set(key, kvHash[:]) // TODO: key or hash(key)?
+	if err != nil {
+		panic(err.Error())
+	}
+}
+
+func (s *Store) Delete(key []byte) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.sc.Delete(key)
+
+	defer func() {
+		_ = s.data.Delete(key)
+	}()
+
+	value, err := s.data.Get(key)
+	if err != nil {
+		panic(err.Error())
+	}
+	kvHash := sha256.Sum256(append(key, value...))
```

Comment:
Instead of looking up the value and hashing it and the key to reconstruct `hash(key, value)` we could lookup the `hash(key, value)` from the SC before deleting it. This appears preferable since both approaches require a DB lookup but this approach doesn't require a hashing step. What do you think?

Reply:
That does indeed make more sense